Source/AwsCommonRuntimeKit/crt/CBOR.swift

```diff
@@ -0,0 +1,360 @@
+            }
+        case .undefined: aws_cbor_encoder_write_undefined(self.rawValue)
+        case .indef_break: aws_cbor_encoder_write_break(self.rawValue)
+        case .indef_array_start: aws_cbor_encoder_write_indef_array_start(self.rawValue)
+        case .indef_map_start: aws_cbor_encoder_write_indef_map_start(self.rawValue)
+        case .indef_bytes_start: aws_cbor_encoder_write_indef_bytes_start(self.rawValue)
+        case .indef_text_start: aws_cbor_encoder_write_indef_text_start(self.rawValue)
+        }
+    }
+
+    /// Get all the values encoded so far as an array of raw bytes.
+    /// This won't reset the encoder, and you will get all the bytes encoded so far from the beginning.
+    public func getEncoded() -> [UInt8] {
+        aws_cbor_encoder_get_encoded_data(self.rawValue).toArray()
+    }
+
+    deinit {
+        aws_cbor_encoder_destroy(rawValue)
+    }
+}
+
+/// Decoder for the CBOR encoding.
+public class CBORDecoder {
+    var rawValue: OpaquePointer
+    // Keep a reference to data to make it outlive the decoder
+    var data: [UInt8]
+
+    public init(data: [UInt8]) throws {
+        self.data = data
+        let count = self.data.count
+        let rawValue = self.data.withUnsafeBytes {
+            let cursor = aws_byte_cursor_from_array($0.baseAddress, count)
+            return aws_cbor_decoder_new(allocator.rawValue, cursor)
+        }
+        guard let rawValue else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+        self.rawValue = rawValue
+    }
+
+    // swiftlint:disable function_body_length 
+    /// Decodes and returns the next value. If there is no value, this function will throw an error. 
+    /// You must call `hasNext()` before calling this function.
+    public func popNext() throws -> CBORType {
+        var cbor_type: aws_cbor_type = AWS_CBOR_TYPE_UNKNOWN
+        guard aws_cbor_decoder_peek_type(self.rawValue, &cbor_type) == AWS_OP_SUCCESS else {
+            throw CommonRunTimeError.crtError(.makeFromLastError())
+        }
+        switch cbor_type {
+        case AWS_CBOR_TYPE_UINT:
+            do {
+                var out_value: UInt64 = 0
+                guard
+                    aws_cbor_decoder_pop_next_unsigned_int_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .uint64(out_value)
+            }
+
+        case AWS_CBOR_TYPE_NEGINT:
+            do {
+                var out_value: UInt64 = 0
+                guard
+                    aws_cbor_decoder_pop_next_negative_int_val(self.rawValue, &out_value)
+                        == AWS_OP_SUCCESS
+                else {
+                    throw CommonRunTimeError.crtError(.makeFromLastError())
+                }
+                return .int(Int64(-Int64(out_value) - 1))
```

Comment:
The cbor negative int supports ` [-2^64; -1]`, it may overflow here. Check for overflow